driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java

```diff
@@ -126,23 +127,43 @@ public ServerDescription getDescription() {
 
     @Override
     public void invalidate() {
+        invalidate(connectionPool.getGeneration());
+    }
+
+    @Override
+    public void invalidate(final int connectionGeneration) {
         if (!isClosed()) {
+            if (connectionGeneration < connectionPool.getGeneration()) {
+                return;
+            }
             serverStateListener.stateChanged(new ChangeEvent<ServerDescription>(description, ServerDescription.builder()
                     .state(CONNECTING)
                     .address(serverId.getAddress())
                     .build()));
             connectionPool.invalidate();
+            // TODO: in streaming protocol, we want to close the current connection and start over
             connect();
         }
     }
 
     @Override
-    public void invalidate(final Throwable t) {
+    public void invalidate(final Throwable reason) {
+        invalidate(reason, connectionPool.getGeneration(), description.getMaxWireVersion());
+    }
+
+    @Override
+    public void invalidate(final Throwable t, final int connectionGeneration, final int maxWireVersion) {
         if (!isClosed()) {
+            if (connectionGeneration < connectionPool.getGeneration()) {
+                return;
+            }
             if ((t instanceof MongoSocketException && !(t instanceof MongoSocketReadTimeoutException))) {
                 invalidate();
             } else if (t instanceof MongoNotPrimaryException || t instanceof MongoNodeIsRecoveringException) {
-                if (description.getMaxWireVersion() < FOUR_DOT_TWO_WIRE_VERSION) {
+                if (isStale(((MongoCommandException) t))) {
```

Comment:
This is the second requirement: that we ignore these particular errors when the topology version is stale.